File: src/services/order.service.ts
import crypto from 'crypto';
import { prisma } from '../config/prisma';
import { AppError } from '../middleware/error';
import { PaymentService } from './payment.service';
import { PaymentMethod, OrderStatus } from '@prisma/client';

export class OrderService {
  public static async checkout(data: {
    cartId: string;
    customerEmail: string;
    customerPhone?: string | null;
    shippingAddress?: string | null;
    paymentMethod: PaymentMethod;
  }) {
    const { cartId, customerEmail, customerPhone, shippingAddress, paymentMethod } = data;

    const cart = await prisma.cart.findUnique({
      where: { id: cartId },
      include: {
        items: {
          include: {
            product: true,
          },
        },
      },
    });

    if (!cart || cart.items.length === 0) {
      throw new AppError('Your cart is empty', 400);
    }

    for (const item of cart.items) {
      if (item.quantity > item.product.stock) {
        throw new AppError(
          `Insufficient stock for ${item.product.name}. Requested: ${item.quantity}, Available: ${item.product.stock}`,
          400
        );
      }
    }

    let totalAmount = 0;
    for (const item of cart.items) {
      totalAmount += Number(item.product.price) * item.quantity;
    }

    const orderNumber = `ORD-${Date.now()}-${crypto.randomBytes(3).toString('hex').toUpperCase()}`;

    // Execute atomic transaction for checkout
    const order = await prisma.$transaction(async (tx) => {
      const newOrder = await tx.order.create({
        data: {
          orderNumber,
          cartId,
          customerEmail,
          customerPhone,
          shippingAddress,
          totalAmount,
          status: OrderStatus.PENDING,
          paymentMethod,
          items: {
            create: cart.items.map((item) => ({
              productId: item.productId,
              quantity: item.quantity,
              price: item.product.price, // Price snapshot
            })),
          },
        },
        include: {
          items: {
            include: {
              product: true,
            },
          },
        },
      });

      // Decrement product inventory
      for (const item of cart.items) {
        await tx.product.update({
          where: { id: item.productId },
          data: {
            stock: {
              decrement: item.quantity,
            },
          },
        });
      }

      // Clear Cart items
      await tx.cartItem.deleteMany({
        where: { cartId },
      });

      return newOrder;
    });

    // Create Payment (COD or Razorpay order)
    const paymentResult = await PaymentService.createOrder({
      amount: Number(order.totalAmount),
      currency: 'INR',
      method: paymentMethod,
      customerEmail,
      customerPhone,
    });

    // Link Payment record to Order
    await prisma.payment.update({
      where: { orderId: paymentResult.orderId },
      data: {
        orderRefId: order.id,
      },
    });

    return {
      orderId: order.id,
      orderNumber: order.orderNumber,
      totalAmount: Number(order.totalAmount),
      status: order.status,
      paymentMethod: order.paymentMethod,
      items: order.items.map((item) => ({
        id: item.id,
        productId: item.productId,
        name: item.product.name,
        price: Number(item.price),
        quantity: item.quantity,
        itemTotal: Number(item.price) * item.quantity,
      })),
      payment: paymentResult,
    };
  }

  public static async getOrderById(id: number) {
    const order = await prisma.order.findUnique({
      where: { id },
      include: {
        items: {
          include: {
            product: true,
          },
        },
        payment: true,
      },
    });

    if (!order) {
      throw new AppError(`Order with ID ${id} not found`, 404);
    }

    return {
      id: order.id,
      orderNumber: order.orderNumber,
      customerEmail: order.customerEmail,
      customerPhone: order.customerPhone,
      shippingAddress: order.shippingAddress,
      totalAmount: Number(order.totalAmount),
      status: order.status,
      paymentMethod: order.paymentMethod,
      createdAt: order.createdAt,
      items: order.items.map((item) => ({
        id: item.id,
        productId: item.productId,
        name: item.product.name,
        price: Number(item.price),
        quantity: item.quantity,
        imageUrl: item.product.imageUrl,
        itemTotal: Number(item.price) * item.quantity,
      })),
      payment: order.payment
        ? {
            orderId: order.payment.orderId,
            paymentId: order.payment.paymentId,
            status: order.payment.status,
            method: order.payment.method,
          }
        : null,
    };
  }
}

File: backend/src/main/java/com/razorpay/integration/service/OrderService.java
package com.razorpay.integration.service;

import com.razorpay.integration.dto.*;
import com.razorpay.integration.exception.AppException;
import com.razorpay.integration.model.*;
import com.razorpay.integration.repository.*;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.security.SecureRandom;
import java.util.ArrayList;
import java.util.HexFormat;
import java.util.List;
import java.util.stream.Collectors;

@Slf4j
@Service
@RequiredArgsConstructor
public class OrderService {

    private final OrderRepository orderRepository;
    private final CartRepository cartRepository;
    private final CartItemRepository cartItemRepository;
    private final ProductRepository productRepository;
    private final UserRepository userRepository;
    private final PaymentRepository paymentRepository;
    private final PaymentService paymentService;
    private final WhatsAppService whatsAppService;

    private static final SecureRandom RANDOM = new SecureRandom();

    @Transactional
    public OrderResponseDto checkout(CheckoutRequest request, Integer userId) {
        Cart cart = cartRepository.findById(request.getCartId())
                .orElseThrow(() -> new AppException("Your cart is empty", HttpStatus.BAD_REQUEST));

        if (cart.getItems() == null || cart.getItems().isEmpty()) {
            throw new AppException("Your cart is empty", HttpStatus.BAD_REQUEST);
        }

        for (CartItem item : cart.getItems()) {
            if (item.getQuantity() > item.getProduct().getStock()) {
                throw new AppException(
                        String.format("Insufficient stock for %s. Requested: %d, Available: %d",
                                item.getProduct().getName(), item.getQuantity(), item.getProduct().getStock()),
                        HttpStatus.BAD_REQUEST
                );
            }
        }

        BigDecimal totalAmount = BigDecimal.ZERO;
        for (CartItem item : cart.getItems()) {
            BigDecimal itemTotal = item.getProduct().getPrice().multiply(BigDecimal.valueOf(item.getQuantity()));
            totalAmount = totalAmount.add(itemTotal);
        }

        byte[] bytes = new byte[3];
        RANDOM.nextBytes(bytes);
        String orderNumber = "ORD-" + System.currentTimeMillis() + "-" + HexFormat.of().formatHex(bytes).toUpperCase();

        User user = userId != null ? userRepository.findById(userId).orElse(null) : null;

        Order order = Order.builder()
                .orderNumber(orderNumber)
                .user(user)
                .cartId(request.getCartId())
                .customerEmail(request.getCustomerEmail())
                .customerCountryCode(request.getCustomerCountryCode() != null ? request.getCustomerCountryCode() : "+91")
                .customerPhone(request.getCustomerPhone())
                .shippingAddress(request.getShippingAddress())
                .totalAmount(totalAmount)
                .status(OrderStatus.PENDING)
                .paymentMethod(request.getPaymentMethod())
                .items(new ArrayList<>())
                .build();

        for (CartItem cartItem : cart.getItems()) {
            OrderItem orderItem = OrderItem.builder()
                    .order(order)
                    .product(cartItem.getProduct())
                    .quantity(cartItem.getQuantity())
                    .price(cartItem.getProduct().getPrice())
                    .build();
            order.getItems().add(orderItem);
        }

        order = orderRepository.save(order);

        // Immediate inventory decrement & cart clear for COD orders
        if (request.getPaymentMethod() == PaymentMethod.COD) {
            for (CartItem cartItem : cart.getItems()) {
                Product product = cartItem.getProduct();
                product.setStock(Math.max(0, product.getStock() - cartItem.getQuantity()));
                productRepository.save(product);
            }
            cartItemRepository.deleteByCartId(request.getCartId());
        }

        // Create Payment record
        CreatePaymentRequest paymentReq = new CreatePaymentRequest();
        paymentReq.setAmount(order.getTotalAmount());
        paymentReq.setCurrency("INR");
        paymentReq.setMethod(request.getPaymentMethod());
        paymentReq.setCustomerEmail(request.getCustomerEmail());
        paymentReq.setCustomerCountryCode(request.getCustomerCountryCode());
        paymentReq.setCustomerPhone(request.getCustomerPhone());

        PaymentResponseDto paymentResult = paymentService.createPayment(paymentReq);

        // Link payment record to Order
        com.razorpay.integration.model.Payment payment = paymentRepository.findByOrderId(paymentResult.getOrderId()).orElse(null);
        if (payment != null) {
            payment.setOrderRef(order);
            paymentRepository.save(payment);
        }

        // Trigger WhatsApp notification for COD
        if (request.getPaymentMethod() == PaymentMethod.COD) {
            String customerName = user != null ? user.getName() : "Customer";
            String toPhone = (request.getCustomerCountryCode() != null ? request.getCustomerCountryCode() : "+91") + request.getCustomerPhone();
            whatsAppService.sendOrderConfirmation(toPhone, customerName, order.getOrderNumber(), order.getTotalAmount(), request.getShippingAddress());
        }

        return OrderResponseDto.builder()
                .orderId(order.getId())
                .orderNumber(order.getOrderNumber())
                .userId(order.getUser() != null ? order.getUser().getId() : null)
                .totalAmount(order.getTotalAmount())
                .status(order.getStatus())
                .paymentMethod(order.getPaymentMethod())
                .payment(paymentResult)
                .build();
    }

    @Transactional(readOnly = true)
    public Order getOrderById(Integer id, Integer userId) {
        Order order = orderRepository.findById(id)
                .orElseThrow(() -> new AppException("Order not found", HttpStatus.NOT_FOUND));

        if (userId != null && (order.getUser() == null || !order.getUser().getId().equals(userId))) {
            throw new AppException("Unauthorized access to this order", HttpStatus.FORBIDDEN);
        }

        return order;
    }

    @Transactional(readOnly = true)
    public OrdersResponseDto getMyOrders(Integer userId, int page, int limit) {
        int pageNumber = Math.max(0, page - 1);
        Pageable pageable = PageRequest.of(pageNumber, limit, Sort.by(Sort.Direction.DESC, "createdAt"));

        Page<Order> orderPage = orderRepository.findByUserId(userId, pageable);

        PaginationDto pagination = PaginationDto.builder()
                .page(page)
                .limit(limit)
                .totalItems(orderPage.getTotalElements())
                .totalPages(orderPage.getTotalPages())
                .build();

        return OrdersResponseDto.builder()
                .orders(orderPage.getContent())
                .pagination(pagination)
                .build();
    }
}
